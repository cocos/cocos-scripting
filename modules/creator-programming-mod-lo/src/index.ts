export * from './mod-lo';
export * from './cjs/share';
