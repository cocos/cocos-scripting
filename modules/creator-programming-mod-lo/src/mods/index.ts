import { CommonJsMod } from "./common-js";
import { EsmMod } from "./esm";
import { JsonMod } from "./json";

export type Mod =
    | EsmMod
    | JsonMod
    | CommonJsMod
    ;
    
export type ModuleType = Mod['type'];

export {
    EsmMod,
    JsonMod,
    CommonJsMod,
};

export {
    JavaScriptSource,
    SourceMap,
    SourceLocation,
    Specifier,
    TransformResolver,
    TransformResult,
} from './common';
