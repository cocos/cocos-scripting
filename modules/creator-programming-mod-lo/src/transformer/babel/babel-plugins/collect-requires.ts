import * as babel from '@babel/core';

export function collectRequires(ast: babel.types.Node) {
    const requires: string[] = [];
    babel.traverse(ast, collectRequiresVisitor(requires));
    return requires;
}

export function collectRequiresVisitor(requires: string[]): babel.Visitor {
    return {
        CallExpression: (path) => {
            const { node } = path;
            if (babel.types.isIdentifier(node.callee) &&
                node.callee.name === 'require' &&
                node.arguments.length === 1 &&
                babel.types.isStringLiteral(node.arguments[0])) {
                requires.push(node.arguments[0].value);
            }
        },
    };
}
