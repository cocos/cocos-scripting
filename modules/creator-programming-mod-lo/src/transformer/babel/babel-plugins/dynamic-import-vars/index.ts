
import type * as babel from '@babel/core';
import decomposeDynamicImportVars from './decompose-dynamic-import-vars';
import type { SpecifierParts } from './decompose-dynamic-import-vars';
import createDefaultResolver from './default-resolve';
import type { Options as DefaultResolveOptions } from './default-resolve';

export type { SpecifierParts };

export type SpecifierCandidate = string | [string, string];

/**
 * Custom specifier resolution.
 * @param specifierParts Specifier parts.
 * @param fileName The source file name(provided by babel).
 * @returns If `undefined` or empty array is returned, this plugin won't do nothing.
 */
export type CustomResolve = (specifierParts: SpecifierParts, fileName: string) => SpecifierCandidate[] | undefined;

export type { DefaultResolveOptions };

export interface Options {
    /**
     * Either the options to configure the default resolver or supply a custom resolver.
     */
    resolve?: DefaultResolveOptions | CustomResolve;
}

export default function({ types }: typeof babel): babel.PluginObj<{
    filename?: string;
    opts?: Options;
}> {
    // Ensures that we won't loop infinitely.
    const myImports: Set<babel.types.Import> = new Set();

    const createImport = () => {
        const importNode = types.import();
        myImports.add(importNode);
        return importNode;
    };

    return {
        visitor: {
            CallExpression: (path, state) => {
                const fileName = state.filename;
                if (!fileName) {
                    return;
                }

                const node = path.node;
                if (node.callee.type !== 'Import' ||
                    node.arguments.length !== 1 ||
                    !types.isExpression(node.arguments[0]) ||
                    types.isStringLiteral(node.arguments[0])) {
                    return;
                }

                if (myImports.has(node.callee)) {
                    return;
                }

                const originalSpecifier = node.arguments[0];

                const specifierParts = decomposeDynamicImportVars(originalSpecifier);
                if (specifierParts.length === 0) {
                    return;
                }

                const options = state.opts ?? {};

                let resolve: CustomResolve;
                if (typeof options.resolve === 'function') {
                    resolve = options.resolve;
                } else {
                    resolve = createDefaultResolver(options.resolve);
                }

                const candidates = resolve(specifierParts, fileName);
                if (!candidates || candidates.length === 0) {
                    return;
                }

                const specifierArg = () => types.identifier('path');

                const candidateSwitchCases = candidates.map((candidate) => {
                    let test: string;
                    let specifier: string;
                    if (Array.isArray(candidate)) {
                        test = candidate[0];
                        specifier = candidate[1];
                    } else {
                        test = candidate;
                        specifier = candidate;
                    }
                    return types.switchCase(
                        types.stringLiteral(test),
                        [types.returnStatement(types.callExpression(
                            createImport(), [types.stringLiteral(specifier)]))],
                    )
                });

                const lambda = types.parenthesizedExpression(
                    types.arrowFunctionExpression(
                        [specifierArg()],
                        types.blockStatement([types.switchStatement(
                            specifierArg(),
                            candidateSwitchCases.concat([types.switchCase(
                                null,
                                [types.returnStatement(types.callExpression(
                                    createImport(), [specifierArg()]))],
                            )]),
                        )])),
                );

                path.replaceWith(types.callExpression(lambda, [ originalSpecifier ]));
            },
        },
    };
}
