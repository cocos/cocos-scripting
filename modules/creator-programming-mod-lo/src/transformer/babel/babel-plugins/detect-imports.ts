
import * as babel from '@babel/core';
import { Specifier } from '../../../mods';

interface LineColumn {
    line: number;
    column: number;
}

interface SourceLocation {
    start: LineColumn;
    end: LineColumn;
}

export function createBabelPluginDetectAndRewriteImports({
    rewrite,
    specifiers,
}: {
    rewrite?: (specifier: string) => string | undefined;
    specifiers: Specifier[];
}) {
    const add = (specifier: string, loc?: babel.types.SourceLocation | null) => {
        const resolved = rewrite?.(specifier);
        specifiers.push({
            value: specifier,
            resolved,
            loc: loc ? {
                start: { line: loc.start.line, column: loc.start.column },
                end: { line: loc.end.line, column: loc.end.column },
            } : undefined,
        });
        return resolved;
    };
    const addStringLiteralSpecifier = (specifier: babel.types.StringLiteral) => {
        return add(specifier.value, specifier.loc);
    };
    const createRewrittenStringLiteralSpecifier = (value: string, original: string) => {
        const stringLiteral = babel.types.stringLiteral(value);
        babel.types.addComment(stringLiteral, 'trailing', original);
        return stringLiteral;
    };
    return {
        visitor: visitImportSpecifiers({
            stringLiteral(path) {
                const rewritten = addStringLiteralSpecifier(path.node);
                if (rewritten) {
                    path.replaceWith(createRewrittenStringLiteralSpecifier(rewritten, path.node.value));
                }
            },
            simpleTemplateLiteral(path, raw) {
                const rewritten = add(raw, path.node.loc);
                if (rewritten) {
                    path.replaceWith(createRewrittenStringLiteralSpecifier(rewritten, raw));
                }
            },
        }),
    };
}

export async function detectImports(ast: babel.types.Program | babel.types.File): Promise<Specifier[]> {
    const result: Specifier[] = [];
    const add = (specifier: string, loc?: babel.types.SourceLocation | null) => {
        result.push({
            value: specifier,
            loc: loc ? {
                start: { line: loc.start.line, column: loc.start.column },
                end: { line: loc.end.line, column: loc.end.column },
            } : undefined,
        });
    };
    const addStringLiteralSpecifier = (specifier: babel.types.StringLiteral) => {
        add(specifier.value, specifier.loc);
    };
    babel.traverse(ast, visitImportSpecifiers({
        stringLiteral(path) {
            addStringLiteralSpecifier(path.node);
        },
        simpleTemplateLiteral(path, raw) {
            add(raw, path.node.loc);
        },
    }));
    return result;
}

interface ImportSpecifierVisitor {
    stringLiteral(path: babel.NodePath<babel.types.StringLiteral>): void;

    simpleTemplateLiteral(path: babel.NodePath<babel.types.TemplateLiteral>, value: string): void;
}

function visitImportSpecifiers<S>(visitor: ImportSpecifierVisitor): babel.Visitor<S> {
    return {
        ImportDeclaration: (path) => visitor.stringLiteral(path.get('source')),
        ExportAllDeclaration: (path) => visitor.stringLiteral(path.get('source')),
        ExportNamedDeclaration: (path) => {
            if (path.node.source) {
                visitor.stringLiteral(path.get('source') as babel.NodePath<babel.types.StringLiteral>);
            }
        },
        CallExpression: (path) => {
            const { callee, arguments: args } = path.node;
            if (callee.type === 'Import' && args.length === 1) {
                const specifierNode = args[0];
                if (specifierNode.type === 'StringLiteral') {
                    visitor.stringLiteral(path.get('arguments')[0] as babel.NodePath<babel.types.StringLiteral>);
                } else if (
                    specifierNode.type === 'TemplateLiteral' &&
                    specifierNode.expressions.length === 0 &&
                    specifierNode.quasis.length === 1) {
                    visitor.simpleTemplateLiteral(
                        path.get('arguments')[0] as babel.NodePath<babel.types.TemplateLiteral>,
                        specifierNode.quasis[0].value.raw,
                    );
                }
            }
        },
    };
}
