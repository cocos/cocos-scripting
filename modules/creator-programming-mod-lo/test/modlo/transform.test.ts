
import { pathToFileURL } from 'url';
import { Mod, ModLo } from '../../src/mod-lo';
import ps from 'path';
import { SpecifierParts } from '../../src/transformer/babel/babel-plugins/dynamic-import-vars';
import * as babel from '@babel/core';
import decomposeDynamicImportVars from '../../src/transformer/babel/babel-plugins/dynamic-import-vars/decompose-dynamic-import-vars';
import createDefaultResolver from '../../src/transformer/babel/babel-plugins/dynamic-import-vars/default-resolve';

describe('Transform CommonJS modules', () => {
    test.each([
        ['ESM', true],
        ['SystemJS', false],
    ])(`%s`, async (_title, isEsm) => {
        const file = pathToFileURL(ps.join(__dirname, 'inputs', 'cjs-es6.js'));
        const modLo = new ModLo({
            _compressUUID: () => '',
            transformExcludes: [
                file.href,
            ],
        });
        const mod = await modLo.load(file);
        const code = isEsm
            ? (await mod.module()).code
            : (await mod.systemjs()).source.code
            ;
        expect(code).toMatchSnapshot(`isEsm: ${isEsm}`);
    });
});

describe('Transform option: _importMetaURLValid', () => {
    test.each([
        [true],
        [false],
    ])(`%s`, async (importMetaURLValid) => {
        const modLo = new ModLo({
            _compressUUID: () => '',
            _importMetaURLValid: importMetaURLValid,
        });
        const url = pathToFileURL(ps.join(__dirname, 'inputs', 'cjs-import-meta-url-valid.js'));
        const mod = await modLo.load(url);
        const code = (await mod.module()).code;
        expect(code).toMatchSnapshot();
    });
});

test('Transform JSON module', async () => {
    const modLo = new ModLo({
        _compressUUID: () => '',
    });
    const file = pathToFileURL(ps.join(__dirname, 'inputs', 'json-module.json'));
    const mod = await modLo.load(file);
    const code = (await toEsm(mod)).code;
    expect(code).toMatchSnapshot();
});

describe('Transform options: useDefineForClassFields', () => {
    test.each([
        [true],
        [false],
    ])(`%s`, async (useDefineForClassFields) => {
        const file = pathToFileURL(ps.join(__dirname, 'inputs', 'use-define-for-class-fields.mjs'));
        const modLo = new ModLo({
            _compressUUID: () => '',
            useDefineForClassFields,
        });
        const mod = await modLo.load(file);
        const code = (await mod.module()).code;
        expect(code).toMatchSnapshot(`${useDefineForClassFields}`);
    });
});

describe('Transform options: useDefineForClassFields & allowDeclareFields', () => {
    test.each([
        [true, true],
        [false, false],
        [true, false],
        [false, true],
    ])(`useDefineForClassFields: %s & allowDeclareFields: %s`, async (useDefineForClassFields, allowDeclareFields) => {
        const file = pathToFileURL(ps.join(__dirname, 'inputs', 'use-define-for-class-fields-and-allow-declare-fields.ts'));
        const modLo = new ModLo({
            _compressUUID: () => '',
            useDefineForClassFields,
            allowDeclareFields,
        });
        const mod = await modLo.load(file);
        const code = (await mod.module()).code;
        expect(code).toMatchSnapshot();
    });
});

describe('Transform options: loose', () => {
    test.each([
        [true],
        [false],
    ])(`%s`, async (loose) => {
        const file = pathToFileURL(ps.join(__dirname, 'inputs', 'loose.mjs'));
        const modLo = new ModLo({
            _compressUUID: () => '',
            loose,
        });
        const mod = await modLo.load(file);
        const code = (await mod.module()).code;
        expect(code).toMatchSnapshot();
    });
});

describe('Transform options: dynamic import vars', () => {
    test('Decompose', async () => {
        await x("import('a.js')", ['a.js']);
        await x("import(`a.js`)", ['a.js']);
        await x("import('a' + '.js')", ['a.js']);
        await x("import(a)", [null]);
        await x("import(`./${a}.js`)", ['./', null, '.js']);
        await x("import(`${a}.js`)", [null, '.js']);
        await x("import(`./${a}`)", ['./', null]);

        async function x(code: string, result: SpecifierParts) {
            const ast = await babel.parseAsync(code, {});
    
            expect(ast).not.toBeNull();
    
            const imports: babel.types.Expression[] = [];
            babel.traverse(ast, {
                CallExpression(path) {
                    if (path.node.callee.type === 'Import') {
                        if (path.node.arguments.length === 1 &&
                            babel.types.isExpression(path.node.arguments[0])) {
                            imports.push(path.node.arguments[0]);
                        }
                    }
                },
            });
            expect(imports.length === 1);
    
            const parts = decomposeDynamicImportVars(imports[0]);
    
            expect(parts).toEqual(expect.arrayContaining(result));;
        }
    });

    test('Default resolve', async () => {
        const fileName = ps.join(__dirname, 'inputs', 'dynamic-import-vars-default-resolve', 'index.js');
        const resolver = createDefaultResolver();
        const resolve = (specifierParts: SpecifierParts) => resolver(specifierParts, fileName);
        
        // Specifiers not starts with './' or '../' are not resolved
        expect(resolve([null])).toBe(undefined);
        expect(resolve([null, '.js'])).toBe(undefined);
        expect(resolve(['.', null,'.js'])).toBe(undefined);
    
        const doubleEach = <T>(arr: T[]) => arr.map((v) => [v, v]);
    
        // Basic
        expect(resolve(['./', null, '.js'])).toEqual(expect.arrayContaining(doubleEach(['./index.js', './1.js', './2.js'])));
        expect(resolve(['./dir/', null, '.js'])).toEqual(expect.arrayContaining(doubleEach(['./dir/1.js', './dir/2.js'])));
    
        // '.ts' can be resolved from '.js'
        expect(resolve(['./ts-dir/', null, '.js'])).toEqual(expect.arrayContaining(doubleEach(['./ts-dir/1.js', './ts-dir/2.js'])));
    });
});

test('Bugfix cocos/cocos-engine#11373', async () => {
    const modLo = new ModLo({
        _compressUUID: () => '',
    });
    const file = pathToFileURL(ps.join(__dirname, 'inputs', 'bugfix#11373', 'assets', 'scripts', 'login', 'AppVersion.js'));
    const mod = await modLo.load(file);
    const code = (await mod.module()).code;
    expect(code).toMatchSnapshot();
});

async function toEsm(mod: Mod) {
    return await mod.module();
}
