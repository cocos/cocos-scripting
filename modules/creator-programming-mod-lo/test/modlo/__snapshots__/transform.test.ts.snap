// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bugfix cocos/cocos-engine#11373 1`] = `
"function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }
var _dec, _class;
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, _toPropertyKey(descriptor.key), descriptor); } }
function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); Object.defineProperty(subClass, "prototype", { writable: false }); if (superClass) _setPrototypeOf(subClass, superClass); }
function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }
function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } else if (call !== void 0) { throw new TypeError("Derived constructors may only return object or undefined"); } return _assertThisInitialized(self); }
function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }
function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }
import * as cc from "cc";
var ccclass = cc._decorator.ccclass;
var AppVersion = (_dec = ccclass('AppVersion'), _dec(_class = /*#__PURE__*/function (_cc$Component) {
  _inherits(AppVersion, _cc$Component);
  var _super = _createSuper(AppVersion);
  function AppVersion() {
    _classCallCheck(this, AppVersion);
    return _super.apply(this, arguments);
  }
  _createClass(AppVersion, [{
    key: "onLoad",
    value: function onLoad() {}
  }]);
  return AppVersion;
}(cc.Component)) || _class);
export { AppVersion as default };"
`;

exports[`Transform CommonJS modules ESM: isEsm: true 1`] = `
"import _cjsLoader0 from 'cce:/internal/ml/cjs-loader.mjs';
import { __cjsMetaURL as _req0} from './foo';
import { __cjsMetaURL as _req2} from 'bar';
let _cjsExports0;
const __cjsMetaURL = import.meta.url;
_cjsLoader0.define(__cjsMetaURL, function (exports, require, module, __filename, __dirname) {
// #region ORIGINAL CODE


 const Foo = require('./foo');
 const Bar = require('bar');

 // These variable names are used accordingly by implementation.
 console.log(
     _req,
     _req1,
     _cjsExports,
     _cjsLoader,
     _Test,
 );

 module.exports = class Test {
     F = new Foo();
     B = new Bar();
 }


// #endregion ORIGINAL CODE

_cjsExports0 = module.exports;


}, () => ({
  './foo': _req0,
  'bar': _req2,
}));
export { _cjsExports0 as default };
export { __cjsMetaURL }
"
`;

exports[`Transform CommonJS modules SystemJS: isEsm: false 1`] = `
"System.register(["cce:/internal/ml/cjs-loader.mjs", "./foo", "bar"], function (_export, _context) {
  "use strict";

  var _cjsLoader0, _req0, _req2, _cjsExports0, __cjsMetaURL;
  _export("default", void 0);
  return {
    setters: [function (_cceInternalMlCjsLoaderMjs) {
      _cjsLoader0 = _cceInternalMlCjsLoaderMjs.default;
    }, function (_foo) {
      _req0 = _foo.__cjsMetaURL;
    }, function (_bar) {
      _req2 = _bar.__cjsMetaURL;
    }],
    execute: function () {
      _export("__cjsMetaURL", __cjsMetaURL = _context.meta.url);
      _cjsLoader0.define(__cjsMetaURL, function (exports, require, module, __filename, __dirname) {
        // #region ORIGINAL CODE

        
         const Foo = require('./foo');
         const Bar = require('bar');

         // These variable names are used accordingly by implementation.
         console.log(
             _req,
             _req1,
             _cjsExports,
             _cjsLoader,
             _Test,
         );

         module.exports = class Test {
             F = new Foo();
             B = new Bar();
         }

;
        // #endregion ORIGINAL CODE

        _export("default", _cjsExports0 = module.exports);
      }, () => ({
        './foo': _req0,
        'bar': _req2
      }));
    }
  };
});"
`;

exports[`Transform JSON module 1`] = `
"export default {
  x: 5,
  y: 6,
  s: 'foo',
  b: true,
  nil: null,
  array: [1, 2, 3, 'string'],
  'some-key': 10
};"
`;

exports[`Transform option: _importMetaURLValid false 1`] = `
"import _cjsLoader from 'cce:/internal/ml/cjs-loader.mjs';
var _cjsExports;
var _foo;
var __cjsMetaURL = '70eb3fd3cd871d41e5f7e819d8bed871-cjs-import-meta-url-valid.js';
_cjsLoader.define(__cjsMetaURL, function (exports, require, module, __filename, __dirname) {
  // #region ORIGINAL CODE

  module.exports.foo = null;

  // #endregion ORIGINAL CODE

  _cjsExports = module.exports;
  _foo = module.exports.foo;
}, {});
export { _cjsExports as default };
export { __cjsMetaURL };"
`;

exports[`Transform option: _importMetaURLValid true 1`] = `
"import _cjsLoader from 'cce:/internal/ml/cjs-loader.mjs';
var _cjsExports;
var _foo;
var __cjsMetaURL = import.meta.url;
_cjsLoader.define(__cjsMetaURL, function (exports, require, module, __filename, __dirname) {
  // #region ORIGINAL CODE

  module.exports.foo = null;

  // #endregion ORIGINAL CODE

  _cjsExports = module.exports;
  _foo = module.exports.foo;
}, {});
export { _cjsExports as default };
export { __cjsMetaURL };"
`;

exports[`Transform options: loose false 1`] = `
"function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }
function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); Object.defineProperty(subClass, "prototype", { writable: false }); if (superClass) _setPrototypeOf(subClass, superClass); }
function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }
function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } else if (call !== void 0) { throw new TypeError("Derived constructors may only return object or undefined"); } return _assertThisInitialized(self); }
function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }
function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
var ValueType = function ValueType() {
  _classCallCheck(this, ValueType);
};
var Vec3 = /*#__PURE__*/function (_ValueType) {
  _inherits(Vec3, _ValueType);
  var _super = _createSuper(Vec3);
  function Vec3() {
    var _this;
    _classCallCheck(this, Vec3);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _super.call.apply(_super, [this].concat(args));
    _defineProperty(_assertThisInitialized(_this), "x", 0.0);
    return _this;
  }
  return Vec3;
}(ValueType);"
`;

exports[`Transform options: loose true 1`] = `
"function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; _setPrototypeOf(subClass, superClass); }
function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }
var ValueType = function ValueType() {};
var Vec3 = /*#__PURE__*/function (_ValueType) {
  _inheritsLoose(Vec3, _ValueType);
  function Vec3() {
    var _this;
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _ValueType.call.apply(_ValueType, [this].concat(args)) || this;
    _this.x = 0.0;
    return _this;
  }
  return Vec3;
}(ValueType);"
`;

exports[`Transform options: useDefineForClassFields & allowDeclareFields useDefineForClassFields: false & allowDeclareFields: false 1`] = `
"function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
var C = function C() {
  _classCallCheck(this, C);
  this.a = 1;
};
export {};"
`;

exports[`Transform options: useDefineForClassFields & allowDeclareFields useDefineForClassFields: false & allowDeclareFields: true 1`] = `
"function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
var C = function C() {
  _classCallCheck(this, C);
  this.a = 1;
  this.b = void 0;
};
export {};"
`;

exports[`Transform options: useDefineForClassFields & allowDeclareFields useDefineForClassFields: true & allowDeclareFields: false 1`] = `
"function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
var C = function C() {
  _classCallCheck(this, C);
  _defineProperty(this, "a", 1);
};
export {};"
`;

exports[`Transform options: useDefineForClassFields & allowDeclareFields useDefineForClassFields: true & allowDeclareFields: true 1`] = `
"function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
var C = function C() {
  _classCallCheck(this, C);
  _defineProperty(this, "a", 1);
  _defineProperty(this, "b", void 0);
};
export {};"
`;

exports[`Transform options: useDefineForClassFields false: false 1`] = `
"function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
var C = function C() {
  _classCallCheck(this, C);
  this.a = 1;
  this.b = void 0;
};"
`;

exports[`Transform options: useDefineForClassFields true: true 1`] = `
"function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
var C = function C() {
  _classCallCheck(this, C);
  _defineProperty(this, "a", 1);
  _defineProperty(this, "b", void 0);
};"
`;
